### PR TITLE
Dynamic cleaning

### PR DESCRIPTION
**Pull request by pawel21**
I have add code to dynamic cleaning in `lstchain_dl1ab.py`
To run script with dynamic cleaning:
`python $script_path -f $input_file -o $output_file -c $config_file --pedestal-cleaning True --dynamic-cleaning True`
What do think about my solution? (@moralejo  , @rlopezcoto  , @jsitarek  )

### CODE REVIEW

File: lstchain/scripts/lstchain_dl1ab.py [hunk 1]
```diff
@@ -68,6 +68,12 @@
                     help='Boolean. True to use pedestal cleaning',
                     default=False)
 
+parser.add_argument('--dynamic-cleaning', action='store',
+                    type=lambda x: bool(strtobool(x)),
+                    dest='dynamic_cleaning',
+                    help='Boolean. True to use dyanmic cleaning',
```
Comment (moralejo) [inline lstchain/scripts/lstchain_dl1ab.py:74]: typo => dynamic

**Comment by moralejo** [comment]
Thanks, @pawel21 

I think however it would not be better to do it this way:

1- make the normal cleaning
2- remove from image pixels which have charge below fraction * mean_3_max_signal

In my view it is more natural to just provide the "minimum fraction of the peak" required for a pixel to survive the cleaning. 

Also, this would "decouple" the settings of the two cleanings; the normal one is for images close to detection threshold, whereas the dynamic thing is to remove tails for bright images (hence I see no need to scale in the same way both picture & boundary).

In the way you did it, the two cleanings are coupled, i.e. if I re-tune the thresholds of the standard cleaning (8,4) to get a better near-threshold performance, this will also change the behaviour for bright images. I prefer to be able to tune the cleaning intended for near-threshold without changing the parameters of bright images, for which I may have already a good data-MC match with whatever settings of the dynamic cleaning. For example, the best-possible "bright image cleaning" should hardly depend on NSB level, but the standard "near-threshold cleaning" will.

**Review by jsitarek — COMMENTED** [review]
Hi @moralejo 

> I think however it would not be better to do it this way:
> 
> 1- make the normal cleaning
> 2- remove from image pixels which have charge below fraction * mean_3_max_signal

I am not sure if it is a good idea to do it this way. You can end up with holes in the tail of the image, and strange shapes due to  single pixels going above/below the threshold. I would expect that the current implementation scaling both thresholds should end up with on average a bit more smooth images. 

about the second comment in using the fraction of the peak signal I agree with you, it is easier to optimize the parameters this way.

File: lstchain/scripts/lstchain_dl1ab.py [hunk 2]
```diff
@@ -68,6 +68,12 @@
                     help='Boolean. True to use pedestal cleaning',
                     default=False)
 
+parser.add_argument('--dynamic-cleaning', action='store',
```
[the remark this answers is not in the document]
Comment (rlopezcoto) [reply lstchain/scripts/lstchain_dl1ab.py:71]: I think the best would be is the activation of this option is done similarly as it is done for the pedestal cleaning

**Comment by moralejo** [comment]
> 
> > I think however it would be better to do it this way:
> > 1- make the normal cleaning
> > 2- remove from image pixels which have charge below fraction * mean_3_max_signal
> 
> I am not sure if it is a good idea to do it this way. You can end up with holes in the tail of the image, and strange shapes due to single pixels going above/below the threshold. I would expect that the current implementation scaling both thresholds should end up with on average a bit more smooth images.
> 

I see no problem with holes in the image. In the standard cleaning we are trying to identify the signal against background fluctuations, so we try to "extend" the core of the image via the boundary  threshold, etc. And we may suspect of disconnected pixels - hence we don't keep isolated pixels which are above picture threshold.

But here we are in a different situation: we already start from a set of pixels of which we think they are, with high probability, signal pixels, and there is no need to "revise" that assessment.  What would be the advantage of not having holes between different parts of the image? The key point is to have the same in data and MC, holes or not...

Your comment made me think anyway that we should re-calculate the island quantities after the dynamic threshold is applied.

**Comment by jsitarek** [comment]
Hi @moralejo 
My main worry was a situation that you have 2 pixels close to the (increased) threshold, one about and another below. They survive the original cleaning together, but then this new condition throws away one of them, but the second, single pixel remains. Same happens with the islands, you can have a small island with a few pixels (reflection of a big shower), and you will clean up most of it with the extra cut, but 1 or two pixels might survive by upfluctuating above the new threshold, and making some patchy images with many holes/single pixels. 
If data and MC agree, then sure, I fully agree with you that this does not matter, but since the whole idea of this cleaning was to improve the data/MC agreement, I would say that if we also remove such leftovers we should be able to match data and MCs with without a need to apply this second threshold very strong.

**Comment by moralejo** [comment]
> My main worry was a situation that you have 2 pixels close to the (increased) threshold, one about and another below. They survive the original cleaning together, but then this new condition throws away one of them, but the second, single pixel remains. Same happens with the islands, you can have a small island with a few pixels (reflection of a big shower), and you will clean up most of it with the extra cut, but 1 or two pixels might survive by upfluctuating above the new threshold, and making some patchy images with many holes/single pixels.
> If data and MC agree, then sure, I fully agree with you that this does not matter, but since the whole idea of this cleaning was to improve the data/MC agreement, I would say that if we also remove such leftovers we should be able to match data and MCs with without a need to apply this second threshold very strong.

I prefer to keep it simple. And I don't see an issue with isolated pixels, I think you are simply importing here the "near-threshold problem", where isolate pixels might be afterpulses, and so we prefer to remove them. But here we are in a regime in which that should not happen, because all the pixels we are considering here already passed the standard cleaning.  So I don't think isolated pixels would be any more damaging than _pairs_ of isolated pixels (which I understand you would accept).

**Comment by jsitarek** [comment]
I fully agree to make it simple, but let me give one more clarification to my reasoning. 

Actually I did not mean afterpulses, but the genuine pixels that are a part of the shower, but are in the tails that we want to remove. 
If the problem is the reflections of the main shower with individual mispointed mirrors that create a small copy of the shower in the other part of the camera, in the scheme that you propose we would need to set the threshold high enough to remove all of them. 

Imagine that the cleaning is 8phe and you have a shower that has a peak at ~2400 phe, with a one mispointing mirror reflecting 1/200 of light i.e. ~12 phe. And imaging that the core of the shower is 3 pixels. At 12 phe, the fluctuations are high, ~4 phe, so you can end up easily with one of those pixels having 8 phe, another 11.5 phe, another 16 phe. If you set the second threshold at 1/200 of the peak, i.e. 12 phe, you will remove 2 pixels, but the third one is left, and even one pixel, if located far away from the main image can spoil the Width somewhat. To remove all 3 we would need much higher second threshold, 16/2400, that would also cut more pixels from the main island (which might be a good or a bad thing). On the other hand if you just raise the cleaning thresholds for this event you would remove the two pixels, and the minimum number of neighbours condition would remove the left over, even if it upfluctuated to survive the original threshold. 

But sure, if you still prefer to keep it simple and just apply a second mask on top of normal cleaning it is also fine with me.

**Comment by moralejo** [comment]
> I fully agree to make it simple, but let me give one more clarification to my reasoning.
> 
> Actually I did not mean afterpulses, but the genuine pixels that are a part of the shower, but are in the tails that we want to remove.
> If the problem is the reflections of the main shower with individual mispointed mirrors that create a small copy of the shower in the other part of the camera, in the scheme that you propose we would need to set the threshold high enough to remove all of them.

Indeed, that is the idea.  

> 
> Imagine that the cleaning is 8phe and you have a shower that has a peak at ~2400 phe, with a one mispointing mirror reflecting 1/200 of light i.e. ~12 phe. And imaging that the core of the shower is 3 pixels. At 12 phe, the fluctuations are high, ~4 phe, so you can end up easily with one of those pixels having 8 phe, another 11.5 phe, another 16 phe. If you set the second threshold at 1/200 of the peak, i.e. 12 phe, you will remove 2 pixels, but the third one is left, and even one pixel, if located far away from the main image can spoil the Width somewhat. To remove all 3 we would need much higher second threshold, 16/2400, that would also cut more pixels from the main island (which might be a good or a bad thing). 

I think groups of purposefully misaligned mirrors won't be the only problem. Tails of the psf are probably not very well simulated either, so I don't see a problem with trimming the main island in the process. 
Of course, all of this has to be tested, but I would start from a really simple approach.

**Comment by pawel21** [comment]
Hi @moralejo ,

> I think however it would be better to do it this way:

> 1- make the normal cleaning

> 2- remove from image pixels which have charge below fraction * mean_3_max_signal

I tried implemented ,,fraction cleaning", but I found out that image afrer fraction cleaning looks worst than after standard cleaning (pedestal tailcut cleaning), number of islands increasing. I am attaching two images and histogram, which show number of islands with standard cleaning and fraction cleaning.
image_fraction_cleaning2
image_fraction_cleaning
hist_cleaning_n_isl

**Comment by moralejo** [comment]
Hi Pawel,

those are events with ~70 and ~90 peak charges. We don't want to alter such events... With 4 pe boundary cleaning threshold,  it would take as much as ~6% of the mirror dish (i.e. 24 m2) to be mis focused _coherently_ on a second spot in order to produce a spurious 4 p.e. signal.   If we are ever in such bad mirror situation we'd rather stop operating and fix the dish.

Surely you have used a too high fraction in those examples.

In any case, there is no strong reason to assume "more islands" = "worse analysis". You may be killing hadronic features with your proposed approach.
The whole core / boundary /neighbors things etc are to identify weak signals through their attachment to brighter parts of the image. That is achieved already with the original cleaning, and there is no need to repeat that process when applying this trick.

**Comment by pawel21** [comment]
> Surely you have used a too high fraction in those examples.
Yes, I have used a too high fraction (0.15)

Now, I am showing images with fraction 0.02.

I will implement this ,,fraction cleaning" in the script .
fraction_cleaning_img1
fraction_cleaning_img2
fraction_cleaning_img3

**Comment by maxnoe** [comment]
@pawel21 How can the fraction cleaning increase the number of islands?

I think the fraction should be an lower bound for the actual cleaning limit, so very similar to the pedestal cleaning, it should do something like this:

```
relative_threshold = fraction * max_pixel
threshold = np.where(pixel_threshold >relative_threshold, pixel_threshold, relative_threshold) 
```

File: lstchain/scripts/lstchain_dl1ab.py [hunk 3]
```diff
@@ -119,6 +125,10 @@ def main():
         picture_th, boundary_th, isolated_pixels, min_n_neighbors = cleaning_params
         log.info(f"Tailcut config used: {config['tailcut']}")
 
+    if args.dynamic_cleaning:
+        THRESHOLD_DYNAMIC_CLEANING = config['dynamic_cleaning']['threshold']
+        log.info(f"Using dynamic cleaning with threshold = {config['dynamic_cleaning']['threshold']}")
```
[the remark this answers is not in the document]
Comment (pawel21) [reply lstchain/scripts/lstchain_dl1ab.py:131]: Done

File: lstchain/scripts/lstchain_dl1ab.py [hunk 4]
```diff
@@ -199,13 +209,22 @@ def main():
                     image = smear_light_in_pixels(image,
                                                   camera_geom,
                                                   smeared_light_fraction)
+                CLEANING_FACTOR = 1
+                if args.dynamic_cleaning:
+                    max_3_value_index = np.argsort(image)[-3:]
```
[the remark this answers is not in the document]
Comment (pawel21) [reply lstchain/scripts/lstchain_dl1ab.py]: I am not sure, that can complicate the script more, and we will usually use  the default value (3) anyway

File: lstchain/scripts/lstchain_dl1ab.py (hunk 1 above)
Comment (pawel21) [reply lstchain/scripts/lstchain_dl1ab.py]: Fixed

**Comment by pawel21** [comment]
I modified script according to @moralejo suggestion.

**Comment by moralejo** [comment]
> > Surely you have used a too high fraction in those examples.
> > Yes, I have used a too high fraction (0.15)
> 
> Now, I am showing images with fraction 0.02.
> 

Thanks! Of the shown events only the last one seems to have the spurious images (the three clusters at more or less the same distance from the image core).  Probably just because it is the brightest of the three. The fake islands are gone with the dynamic cleaning, but so is what might be a genuine part of the shower...  This might impact g/h separation for these very large events (more than the main-island approach?). On the other hand, for dimmer images  it may be the dynamic cleaning (which will do nothing to them) is less harming than the main-island approach.  That effect should perhaps be tested on pure MC.

**Review by moralejo — APPROVED** [review]

**Pull request merged**